### DISCUSSION TITLE
open_agb_firm output lag test. Tell me which one has more lag.

### DISCUSSION THREAD

**Discussion by profi200** [profi200/open_agb_firm#135] [Polls]
Since this keeps coming up i thought let's make a little blind test. I provide 2 builds here and one of them has more output lag than the other but i won't tell you which (yet).

Ideally test this with games where lag does matter like rhythm games or fast Jump & runs.

Happy testing.
[open_agb_firm_lag_test_1.zip](https://github.com/profi200/open_agb_firm/files/12702899/open_agb_firm_lag_test_1.zip)
[open_agb_firm_lag_test_2.zip](https://github.com/profi200/open_agb_firm/files/12702900/open_agb_firm_lag_test_2.zip)

**Comment by branchus** [comment]
To me. I cannot tell the difference. Probably I mostly play JRPG/Strategy games and latency is not a big deal in those kind of games.

I like this little app. turned my 3ds into a ultimate handhold.

**Reply by profi200** [reply]
Yeah, with these types of games you will not notice the difference. The idea was to test the theory that normal people can tolerate 2 frames of lag. Normal builds have 1 frame + 1.6 ms of lag if my calculations are not wrong. I have no equipment to actually test this unfortunately.

I wanted to see if it actually matters and if not i may have more room to add enhancements like actual [proper color correction](https://gc-forever.com/wiki/index.php?title=Game_Boy_Interface/Standard_Edition#Color_matrix). Due to hardware limitations this is actually quite difficult to do without sacrificing on other ends.

**Reply by branchus** [reply]
Really happy to see you are actively working on this. Yes, I know it is working on bare metal and add functions are not that easy. Before you add new functions, can you release another stable release please?

**Reply by twocows** [reply]
I think most people won't be able to tell the difference but someone who's a speedrunner or something will be able to because they've trained themselves to... lots of tricks have a one or two frame window so I can see it being a thing for that specific case. Vast majority of people probably won't notice, though.

**Reply by BrunoNRS** [reply]
I think you could something like the gbi for the gameboy player for the GameCube a super low latency version that does not support any advanced features and a normal version with more customization and features.

**Reply by Contra-S** [reply]
> I wanted to see if it actually matters and if not i may have more room to add enhancements like actual [proper color correction](https://gc-forever.com/wiki/index.php?title=Game_Boy_Interface/Standard_Edition#Color_matrix). Due to hardware limitations this is actually quite difficult to do without sacrificing on other ends.

I was wondering if it'd be possible to just patch the color correction into the roms to avoid sacrificing latency, as some Color Restoration romhacks do for games ported to GBA, like this one for [Yoshi's Island](https://www.romhacking.net/hacks/2597/).

**Reply by profi200** [reply]
ROM patches are already supported. Name the patch the same as the ROM (except for the file extension) and place it in the same folder as the ROM. Color correction is not gonna replace color restoration hacks.

**Comment by branchus** [comment]
Please keep an option toggle for this enhancement for those who opt in or those who prefer lower latency.

**Reply by profi200** [reply]
I don't plan on enabling color correction permanently.

**Comment by NintendoManiac64** [comment]
I don't know if I just found a silver bullet test case that no one else has or if I'm just contrarian but, when using DDR GB 2 running through Goomba Color on my O3DS, I'm consistently getting much better scores with build 1 implying that build 2 has more lag (I had a perfect combo on build 1 before switching to build 2 at which point I would always end up with some random misses and/or a broken combo, and then as soon as I switched back to build 1 I immediately got another perfect combo).

**EDIT**: Unless Goomba Color itself introduces some audio delay, thereby any slight input lag might actually be *improving* the audio-video sync? In which case this test of mine would actually be proving the exact opposite—that being build 1 would have more lag.

Regardless, I put my vote in as build 2 is the laggier one.

　
If you want to try this out yourself, I've attached an IPS patch that automatically turns DDR GB 2 into a GBA rom file via Goomba Color (you may need to manually append a .gba file extension to the resulting patched ROM) and I've also attached a custom control configuration that should make your life much easier (makes the ABXY map to a second copy of the normal D-pad directions like is used for the PlayStation versions of the DDR games as well as various homebrew StepMania ports／clones; you can also use L as a "Back／Cancel" button and R as a "Confirm／OK" button):

- [DDR GB 2 to GBA.zip](https://github.com/profi200/open_agb_firm/files/12900309/DDR.GB.2.to.GBA.zip)

Protips and recommendations:
1. On the title screen, don't press A or you'll mess up the controls—press the actual Start button instead
2. On the main menu, ideally select "Free play mode"
3. I personally recommend testing with the song "Mr. Wonderful" since it essentially has no simultaneous double-notes which greatly helps to avoid issue #130, but other songs may work just as well
4. On the song selection screen, double-press down to increase difficulty and double-press up to decrease difficulty

　
————————————————————————————————

　
> I like this little app. turned my 3ds into a ultimate handhold.

@branchus
At the risk of being off-topic, may I ask what your solution is for Game Boy／Color games? I imagine you're using open_agb_firm because mGBA is too slow for you which implies to me that you're using an O3DS (original 3DS) which is what I'm also using along with an O2DS (original 2DS) that my sister has.

Unless I'm missing something obvious, I can't help but notice that the O3DS／O2DS seems to have no ideal solution for Game Boy／Color games outside of the not-exactly convenient VC injection method:
- mGBA and GameYob 3DS have slow-down (the main menu of Pokemon Puzzle Challenge is a good example)
- DSi version of GameYob actually performs better than the 3DS version but has bad upscaling due to the double-scaling from 144p ▶ 192p ▶ 240p
- Goomba／Color running on open_agb_firm can skip frames (going down a warp pipe in the first level of Super Mario Land 2: 6 Golden Coins is a good example)

I can't help but wonder if this hasn't been solved because N3DS／N2DS (new 3DS／new 2DS) have 3x the CPU clockspeed which automatically brute-forces through the performance issues that mGBA has, thereby avoiding this issue altogether.

**Reply by branchus** [reply]
Hi @NintendoManiac64 

I cannot really answer your question. I don't play GB and GBC games so not really thought about that, sorry. 

I have N3DS, but I never used mGBA on it. I use Open AGB because it run GBA game natively on 3DS. It is a personal preference, I know probably a little bit wired, I always prefer native > cycle accurate > LLE emulator > HLE emulator

I know mGBA is quite accurate, and provided a lot of additional functions, I may use it on PC, but on 3DS I still prefer to run game natively.

**Reply by BrunoNRS** [reply]
For gb/gbc ia use a mix of gambatte and mgba on RetroArch, gambatte for normal games and mgba for super Gameboy games, all have border support I think they run great with zero slow down

**Reply by NintendoManiac64** [reply]
You do admittedly get substantially more battery life running GBA games natively...

>  I think they run great with zero slow down

Many games are generally fine but, as mentioned, there are definitely those that aren't with the main menu of Pokemon Puzzle Challenge being the most extreme example I've found to date where it runs at nearly half-speed.

Unless you meant for Gambatte? Last time I tried RetroArch on my O3DS around 2-3 years ago I found it to have kind of ridiculous load times for a device that uses solid state storage, like upwards of 1 minute which automatically made it a non-starter for me (I never grew up with CD-based consoles, and the GameCube had surprisingly minimal load times)

**Reply by BrunoNRS** [reply]
Oh boy, you need to try it again, you're on a o3ds so I don't think the mgba core is gonna be great, but RetroArch loads fast and the gambatte core is great specialty with overlay and speed up options

**Reply by NintendoManiac64** [reply]
>  I don't think the mgba core is gonna be great

Just to clarify, I was talking about stand-alone mGBA which is available for the 3DS as a cia or 3dsx.

**Reply by BrunoNRS** [reply]
Yes I know,  mgba standalone is much faster for GBA games yes, but for gb/gbc the speed is mostly the same, and while I prefer gambatte for gb and gbc games mgba support for sgb enhanced games is much more mature 
And RetroArch has extra features like overlays and better image scaling manipulation so I prefer it over mgba standalone or gameyob

**Comment by profi200** [comment]
Update:
This thread/poll is no longer relevant. I managed to reach my goal without affecting output latency. See [this commit](https://github.com/profi200/open_agb_firm/commit/bf079b7c7ff10e3175f106743c57ebed236afa88).

Poll results at the time of writing this comment. 14 votes.
First build. 35%
Second build. 21%
I can't tell the difference. 42%

open_agb_firm_lag_test_1.zip was set to ~16 ms of extra lag.
open_agb_firm_lag_test_2.zip was the same as regular builds with ~1.6 ms.
